### DISCUSSION TITLE
Third-party external integrations and domain

### DISCUSSION THREAD

**Discussion by TwiceII** [fraktalio/fmodel#215] [Q&A] [answered]
Hey! First of all, thanks for a great lib and especially for a great blog -- it's one of the clearest and thorough description of this topic I've ever seen.

My question is - how do we reconcile our pristine, pure FP domain core with real-life requirements of interactions with external/third-party services, sending notifications and other side-effecty things?

I know that usually we move these things to the "outside" circles, but I can't find exact place where to put it in your Fmodel. Should they belong to Application service?

After reading [Event Modeling](https://eventmodeling.org/posts/what-is-event-modeling/#automation) Modelling blog, I stumbled on this:

> Our system is going to need to communicate with external services. When the guests in our hotel are paying for their stay when they check out, our system makes a call to a payment processor. We can make the concept of how this occurs with the idea of a “todo list” for some processor in our system. This todo list shows tasks we need to complete. Our processor goes through that list from time to time (could be milliseconds or days) and sends out a command to the external system to process the payment, as an example. The reply from the external system is then translated into an event that we store back in our system. This way we keep the building blocks that we use in our system as something that’s meaningful to us.

Here it seems like call to external system is done directly inside command handler (`Decider` in FModel?). Should we pass some interface inside Decider and then mock it in our domain tests? Kinda ruins FP goodiness for me.

I'd appreciate if you share how you approach this kind of issues in your practical implementations, thank you in advance!

**Comment by idugalic** [comment, marked as answer]
Hi Twicell,

The short answer is that Decider, View, and Saga should stay pure / without side effects of any kind.

SagaManager should be responsible for integrating:

- two aggregates
- aggregate and third-party service
- kafka/rabbitMq listener and the aggregate

SagaManager is acting as an anti-corruption layer in this case, translating:

- events of one aggregate into command of the second aggregate
- events of the aggregate into HTTP service calls
- messages/responses from Kafka/RabbitMQ into commands of the aggregate. 

In this way, the integration is minimizing the runtime coupling between two systems.
Observe how that response/message from the third-party system is always translated into a Command of your own system.
You never skip that input/command, and let that message/response leak deep into your domain language / SagaManager is acting as an anti-corruption layer.

SagaManager is acting as a translator in the event modeling workshop format. 

**I plan to write a blog post on this topic very soon and show you the options you have. **

BTW, you can also override the application layer Aggregate and make it use/compose the third-party service. But, there is a reason we have not promoted that by default in our API, more on that in the blog post...

For example, you can control how you instantiate your object (eg. [EventSourcingAggregate](https://github.com/fraktalio/fmodel/blob/117ce281c894dd458a1902d5244ea81fa8d95037/application/src/commonMain/kotlin/com/fraktalio/fmodel/application/EventSourcingAggregate.kt#L180C1-L187C1))

This is what you have now:
```kotlin
fun <C, S, E> EventSourcingAggregate(
    decider: IDecider<C, S, E>,
    eventRepository: EventRepository<C, E>
): EventSourcingAggregate<C, S, E> =
    object : EventSourcingAggregate<C, S, E>,
        EventRepository<C, E> by eventRepository,
        IDecider<C, S, E> by decider {}
```

You can create your own interface by extending/composing EventSourcingAggregate and your client interface/behaviour:
`MyEventSourcingAggregate<C, S, E>: EventSourcingAggregate<C, S, E>, MyThirdPartyServiceClient`

and a factory/constructor function:
```kotlin
fun <C, S, E> MyEventSourcingAggregate(
    decider: IDecider<C, S, E>,
    eventRepository: EventRepository<C, E>,
   myClient: MyThirdPartyServiceClient
): MyEventSourcingAggregate<C, S, E> =
    object : EventSourcingAggregate<C, S, E>,
        EventRepository<C, E> by eventRepository,
        IDecider<C, S, E> by decider,
        MyThirdPartyServiceClient by myClient {}
```

and compose the handle method (by using `my client` now.) as you find fit. This is [our example](https://github.com/fraktalio/fmodel/blob/117ce281c894dd458a1902d5244ea81fa8d95037/application-vanilla/src/commonMain/kotlin/com/fraktalio/fmodel/application/EventSourcingAggregateExtension.kt#L30C1-L34C16)

In your case you can choose to use that client now:
```kotlin
fun <C, S, E> MyEventSourcingAggregate<C, S, E>.handle(command: C): Flow<E> {
  
   return command
        .fetchEvents()
        .computeNewEvents(command)
        .save()

   val 3partyResult = callMethodsFrom`MyThirdPartyServiceClient`(...)
}
```

Best and thanks for your questions!

**Reply by TwiceII** [reply]
Thanks, Ivan, I'm looking forward to read your blog post about this topic!

**Reply by TwiceII** [reply]
Oh, one more thing -- in your model, Sagas are stateless functions: output commands/actions depend only on input events. Which at first glance seems kinda awkward, don't we need more data from state to produce explicit, "fat" commands? 
In almost every other resources, sagas are described as stateful, and while reading "Event modeling" I assumed that automation pattern allows getting data from various Read Models before creating appropriate commands.

In your practice, do you think delegating the "smartness" to command handlers cover all possible scenarios when we need Sagas/Integrations?

**Reply by idugalic** [reply]
I believe that Sagas should be stateless, and Process Managers should be stateful.  By keeping the Saga stateless you are **radically** forcing all state management (and decision-making) to the EventSourcingAggregate.  This way, you are steering your design into a better place :): **`smart endpoints and dumb pipes`**. Aggregates are endpoints, and Sagas are pipes. 

For example, once your aggregate publishes the event `OrderDelivered`, Saga can react to it and publish the command (or send an HTTP request ) to the Payment provider (Stripe, PayPal,...). The SagaManager is resilient, you can retry/replay that event over and over again in case of payment errors. 

EventModeling promotes another variant of this integration as `automation`.  It is using a MaterializedView rather than a stateless SagaManager, to accumulate events into some Read/View state first, and then publish messages/integrate to/with Payment (by updating that state as success payment or error payment at the end / TODO). With this approach, it is your responsibility to implement such Materialized Views that will serve as a TODO list. I have not found the need to maintain this State and prefer delegating the "smartness" to command handlers/aggregates/deciders. This way, all important state is kept as events and it is not distributed across the system (with stateful Sagas). Backup is easy, Disaster recovery is simpler...

It will be discussed in the blog post. 

Best

**Reply by idugalic** [reply]
I just remembered that Fmodel had ProcessManager modeled two years ago, but we have removed it in favor of Decider(Aggregate) + Saga(SagaManager) = ProcessManager

https://github.com/fraktalio/fmodel/blob/e4f4d5c030fe94a76bc23e46e62eb23f76e8bca8/domain/src/main/kotlin/com/fraktalio/fmodel/domain/Process.kt

**Reply by idugalic** [reply]
Have a look at this Miro https://miro.com/app/board/o9J_ktrZ1rU=/. These three systems are integrated with stateless Sagas.

**Reply by TwiceII** [reply]
Thanks again, Ivan! I really appreciate your detailed answers, can't wait for your blog post!

**Reply by idugalic** [reply]
https://fraktalio.com/blog/integration-of-event-driven-systems.html

**Reply by idugalic** [reply]
Check the [PaymentSagaManager](https://github.com/fraktalio/fmodel-ktor-demo/blob/main/src/main/kotlin/com/fraktalio/application/SagaManager.kt) and how it is registered in the [main application class](https://github.com/fraktalio/fmodel-ktor-demo/blob/main/src/main/kotlin/com/fraktalio/Application.kt)
